lib/location/method_gnss.c

```diff
@@ -370,93 +391,126 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
-		return;
-	}
-
 	if (IS_ENABLED(CONFIG_NRF_CLOUD_PGPS)) {
-		/* ephemerides come from P-GPS; almanacs not desired in this configuration */
+		/* Ephemerides come from P-GPS; almanacs not needed in this configuration. */
 		pgps_agps_request.sv_mask_ephe = agps_request.sv_mask_ephe;
+		pgps_agps_request.data_flags = agps_request.data_flags;
 		agps_request.sv_mask_ephe = 0;
 		agps_request.sv_mask_alm = 0;
-		LOG_DBG("P-GPS request from modem (ephe: 0x%08x)", pgps_agps_request.sv_mask_ephe);
+
+		if (IS_ENABLED(CONFIG_NRF_CLOUD_AGPS)) {
+			/* Time and location come from A-GPS. */
+			pgps_agps_request.data_flags = 0;
+		}
+
+		LOG_DBG("P-GPS request sv_mask_ephe: 0x%08x, data_flags 0x%02x",
+			pgps_agps_request.sv_mask_ephe, pgps_agps_request.data_flags);
 	}
 
-	LOG_DBG("A-GPS request from modem (ephe: 0x%08x alm: 0x%08x flags: 0x%02x)",
-		agps_request.sv_mask_ephe,
-		agps_request.sv_mask_alm,
-		agps_request.data_flags);
 #if defined(CONFIG_NRF_CLOUD_AGPS)
+	if (agps_request.data_flags != 0) {
+		/* With A-GPS, if any of the flags in the data_flags field is set, it's
+		 * advantageous to request everything at the same time because the amount of
+		 * data is so small.
+		 */
+		agps_request.data_flags =
+			NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST |
+			NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST |
+			NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST |
+			NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST |
+			NRF_MODEM_GNSS_AGPS_POSITION_REQUEST;
+	}
+
+	LOG_DBG("A-GPS request sv_mask_ephe: 0x%08x, sv_mask_alm: 0x%08x, data_flags: 0x%02x)",
+		agps_request.sv_mask_ephe, agps_request.sv_mask_alm, agps_request.data_flags);
+
 	/* Check the request. If no A-GPS data types except ephemeris or almanac are requested,
-	 * jump to P-GPS (if enabled)
+	 * jump to P-GPS (if enabled).
 	 */
 	if (method_gnss_agps_required(&agps_request)) {
+		enum lte_lc_nw_reg_status reg_status = LTE_LC_NW_REG_NOT_REGISTERED;
+
+		lte_lc_nw_reg_status_get(&reg_status);
+		if (reg_status == LTE_LC_NW_REG_REGISTERED_HOME ||
+		    reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING) {
+			/* Only store the timestamp when LTE is connected, otherwise it's not
+			 * very likely that the A-GPS request succeeds.
+			 */
+			agps_req_timestamp = k_uptime_get();
+		}
 #if defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
-		k_work_submit_to_queue(location_core_work_queue_get(), &method_gnss_agps_ext_work);
+		location_core_event_cb_agps_request(&agps_request);
 #else
-		k_work_submit_to_queue(
-			location_core_work_queue_get(),
-			&method_gnss_agps_request_work);
+		method_gnss_nrf_cloud_agps_request();
 #endif
 	}
-#endif /* CONFIG_NRF_CLOUD_AGPS */
 #if defined(CONFIG_NRF_CLOUD_PGPS)
+	else if (pgps_agps_request.sv_mask_ephe != 0) {
+		/* When both A-GPS and P-GPS are enabled, this needs to be called only when no
+		 * A-GPS data is requested. When A-GPS data is requested, the nRF Cloud library
+		 * calls the function internally after processing the A-GPS data.
```

Comment:
Nasty dependency to nRF Cloud functionality. How does this work with EXTERNAL features? e.g. ATv2 is calling for nrf_cloud_pgps_notify_prediction()

Reply:
Yes, it took me some time to track down why P-GPS injected the same predictions several times in a row. I think the reasoning is that the application can't do this, because it may not anymore know when A-GPS data has been received and injected.

External works in the same way, it uses the same nRF Cloud library A-GPS and P-GPS implementation.

Reply:
ok, and we cannot run this code here with external?

Reply:
That code is used also with external, right? 😄

Reply:
yes, that how it looked. So, I was confused why ATv2 (which is using location api external config) is also calling for nrf_cloud_pgps_notify_prediction().

Reply:
The ATv2 P-GPS implementation should be checked, there's probably some fine tuning to be done there as well.

Reply:
ATv2 P-GPS is not fully cleaned after taking Location library into use and there's a ticket to check exactly these "duplicate" actions, which I didn't dare to remove originally.